### PR TITLE
TMS-3007 kontrol: gracefully handle key updates

### PR DESCRIPTION
**Pull request by rjeczalik**

### CODE REVIEW

File: sockjsclient/xhr.go [hunk 1]
```diff
@@ -71,6 +73,10 @@ func (x *XHRSession) ID() string {
 }
 
 func (x *XHRSession) Recv() (string, error) {
+	type requestCanceler interface {
```
Comment (cihangir) [inline sockjsclient/xhr.go]: 💯

File: kontrol/handlers.go [hunk 2]
```diff
@@ -82,66 +83,73 @@ func (k *Kontrol) HandleRegister(r *kite.Request) (interface{}, error) {
 	every := onceevery.New(UpdateInterval)
 
 	ping := make(chan struct{}, 1)
-	closed := false
+	closed := int32(0)
+
+	kiteCopy := remote.Kite
```
Comment (cihangir) [inline kontrol/handlers.go]: i couldnt get why did you clone it

Comment (rjeczalik) [inline kontrol/handlers.go]: remote.Kite.Username is mutated at [line 141](https://github.com/koding/kite/pull/155/files#diff-02312926bbc7db566f479305ba8bb9f4R141) and read in the `updaterFunc` and `heartbeatArgs` goroutines - race. The copying makes the goroutines use their own copy - to fix the race.

File: sockjsclient/xhr.go [hunk 3]
```diff
@@ -92,61 +86,84 @@ func (x *XHRSession) Recv() (string, error) {
...
-		defer resp.Body.Close()
 
-		if resp.StatusCode != http.StatusOK {
-			return "", fmt.Errorf("Receiving data failed. Want: %d Got: %d",
-				http.StatusOK, resp.StatusCode)
-		}
+		req.Header.Set("Content-Type", "text/plain")
+
+		select {
+		case <-x.abort:
```
Comment (cihangir) [inline sockjsclient/xhr.go]: i was going to suggest to use ctxhttp but it seems they removed go1.4 support from their upstream

File: kontrol/handlers.go (hunk 2 above)
Comment (cihangir) [inline kontrol/handlers.go]: 👌

File: sockjsclient/xhr.go (hunk 3 above)
Comment (rjeczalik) [inline sockjsclient/xhr.go]: definitely context here would be more idiotmatic 👍

**Pull request merged**